### PR TITLE
feat: Add Sink.exists operator

### PR DESCRIPTION
**Pull request by GreyPlane**
Motivation:
Add the Sink.exists operator. 

co-authored with: He-Pin

Result:
Rich operators set and close: https://github.com/apache/incubator-pekko/issues/973
image

### CODE REVIEW

**Comment by laglangyue** [comment]
add test for javadsl, add document

**Comment by He-Pin** [comment]
First:
Thanks and welcome contributions from China user group.

What still missing, the `exists.md` , the JavaTest, the Demo code, which is needed in the `exists.md`. and in `exists.md`, will need an urlencoded for the scaladoc function signature.

For formating, you can try `scala-cli fmt` , which is fast.

File: docs/src/test/java/jdocs/stream/operators/flow/Exists.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+                .via(Flow.<String>create().exists(isAnomaly))
+                .runForeach(new Procedure<Boolean>() {
+                    @Override
+                    public void apply(Boolean param) throws Exception {
+                        System.out.println("Anomaly detected!");
+                    }
+                }, system);
+        // expected prints:
+        // Anomaly detected!
+        // #exists
```
Comment (He-Pin) [inline docs/src/test/java/jdocs/stream/operators/flow/Exists.java:59]: need format with `javafmtAll`?

File: stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala [hunk 2]
```diff
@@ -1527,6 +1527,28 @@ trait FlowOps[+Out, +Mat] {
...
+   *
+   * '''Emits when''' predicate returned true or upstream completes
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' predicate returned true or upstream completes
+   *
+   * '''Cancels when''' predicate returned true or downstream cancels
+   */
+  def exists(p: Out => Boolean): Repr[Boolean] = map(p).takeWhile(!_, inclusive = true).fold(true) {
```
Comment (He-Pin) [inline stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala:1548]: `map` and `takeWhile` can be expressed with `collectWhile`/`collectFirst`?

File: docs/src/test/java/jdocs/stream/operators/flow/Exists.java (hunk 1 above)
Comment (GreyPlane) [reply docs/src/test/java/jdocs/stream/operators/flow/Exists.java:59]: Thanks for point it out, I didn't know that is a thing.

File: stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala [hunk 3]
```diff
@@ -1527,6 +1527,28 @@ trait FlowOps[+Out, +Mat] {
...
+    case (_, incoming) => incoming
```
Comment (He-Pin) [inline stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala:1549]: need a `name("exists")` for this subFlow.

File: stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala (hunk 2 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala:1548]: Sorry, the `collectFirst` is not merged yet.refs : https://github.com/apache/incubator-pekko/pull/984

Comment (GreyPlane) [reply stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala:1548]: Can't see obvious reason for use `collectWhile` I guess? `inclusive` makes `takeWhile` appropriate for reaching the semantic demand.

File: docs/src/main/paradox/stream/operators/Flow/exists.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# exists
+
+Emit true and complete the stream as soon as there's an element satisfy the *predicate*, or emit false when upstream
```
Comment (He-Pin) [inline docs/src/main/paradox/stream/operators/Flow/exists.md:3]: Emits ,complete.

File: docs/src/main/paradox/stream/operators/Flow/exists.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+@ref[Simple operators](../index.md#simple-operators)
+
+## Signature
+
+@apidoc[Flow.exists](Flow$) { scala="#exists[I](p:Out%20=%3E%20Boolean)" java="#exists[I](p:org.apache.pekko.japi.function.Predicate[I])" }
+
+## Description
+
+Emit true and complete the stream as soon as there's an element satisfy the *predicate*, or emit false when upstream
+complete and no element satisfied the *predicate*.
```
Comment (He-Pin) [inline docs/src/main/paradox/stream/operators/Flow/exists.md:15]: Emits completes

File: docs/src/main/paradox/stream/operators/Flow/exists.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+## Examples
+
+Scala
+:  @@snip [Exists.scala](/docs/src/test/scala/docs/stream/operators/flow/DiMap.scala) { #imports #exists }
+
+Java
+:  @@snip [Exists.java](/docs/src/test/java/jdocs/stream/operators/flow/DiMap.java) { #imports #exists }
```
Comment (He-Pin) [inline docs/src/main/paradox/stream/operators/Flow/exists.md:23]: Another file instead of Dimap.java？

File: docs/src/main/paradox/stream/operators/Flow/exists.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+## Reactive Streams semantics
+
+@@@div { .callout }
+
+**emits** when predicate returned true or upstream completes
+
+**backpressures** '''Backpressures when'' downstream backpressures
+
+**completes** when predicate returned true or upstream completes
```
Comment (He-Pin) [inline docs/src/main/paradox/stream/operators/Flow/exists.md:33]: Return false

File: docs/src/test/scala/docs/stream/operators/sink/Exists.scala [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+  def detectAnomaly(): Unit = {
+    // #exists
+    val source = Source(Seq("Sun is shining", "Unidentified Object", "River is flowing"))
+
+    val anomalies = Seq("Unidentified Object")
+    def isAnomaly(phenomenon: String): Boolean = anomalies.contains(phenomenon)
+
+    val result = source.runWith(Sink.exists(isAnomaly))
+    result.map(println)
```
Comment (He-Pin) [inline docs/src/test/scala/docs/stream/operators/sink/Exists.scala:40]: May need await? The thread may.exsits so quickly

File: stream-tests/src/test/java/org/apache/pekko/stream/javadsl/FlowTest.java [hunk 9]
```diff
@@ -306,6 +306,28 @@ public boolean test(Integer elem) {
...
+                  }
+                });
+
+    final CompletionStage<Done> future =
+        source
+            .via(flow)
+            .runWith(Sink.foreach(elem -> probe.getRef().tell(elem, ActorRef.noSender())), system);
+
+    probe.expectMsgEquals(true);
+    future.toCompletableFuture().get(3, TimeUnit.SECONDS);
```
Comment (He-Pin) [inline stream-tests/src/test/java/org/apache/pekko/stream/javadsl/FlowTest.java:328]: Need an assertions

File: stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowExistsSpec.scala [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.pekko.stream.testkit.scaladsl._
+import org.scalatest.concurrent.ScalaFutures
+
+class FlowExistsSpec extends StreamSpec with ScalaFutures {
+
+  "An Exists" must {
+
+    def greaterThan3(i: Int) = i > 3
+
+    "emit true for empty Source" in {
```
Comment (He-Pin) [inline stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowExistsSpec.scala:30]: false？

Comment (GreyPlane) [reply stream-tests/src/test/scala/org/apache/pekko/stream/scaladsl/FlowExistsSpec.scala]: After talking to fs2 people on discord, it seems like that fs2 returns false was a historical mistake of implementation restriction.
So yes, we should change our implementation to return false for empty stream.

**Comment by GreyPlane** [comment]
docs was removed for now, in favor to fire another PR for adding Sink-or-Flow category for document.

File: stream-tests/src/test/scala/org/apache/pekko/stream/DslConsistencySpec.scala [hunk 11]
```diff
@@ -70,6 +70,8 @@ class DslConsistencySpec extends AnyWordSpec with Matchers {
       "andThenMat",
       "isIdentity",
       "withAttributes",
+      // We box Scala Boolean to Java Boolean for better user experience
+      "exists",
```
Comment (He-Pin) [inline stream-tests/src/test/scala/org/apache/pekko/stream/DslConsistencySpec.scala:74]: Good!

File: stream-tests/src/test/scala/org/apache/pekko/stream/DslFactoriesConsistencySpec.scala [hunk 12]
```diff
@@ -41,8 +41,9 @@ class DslFactoriesConsistencySpec extends AnyWordSpec with Matchers {
       "futureFlow", // completionStageFlow
       "futureSink", // completionStageSink
       "lazyFutureSink", // lazyCompletionStageSink
-      "createGraph" // renamed/overload of create for getting type inference working in Scala 3
-    )
+      "createGraph", // renamed/overload of create for getting type inference working in Scala 3
+      // We box Scala Boolean to Java Boolean for better user experience
+      "exists")
```
Comment (He-Pin) [inline stream-tests/src/test/scala/org/apache/pekko/stream/DslFactoriesConsistencySpec.scala:46]: Can this be fixed with a mapping in `scala -> java types`

Comment (GreyPlane) [reply stream-tests/src/test/scala/org/apache/pekko/stream/DslFactoriesConsistencySpec.scala:46]: Sure, but it feels like a separate PR just like these works on doc IMO

**Comment by He-Pin** [comment]
Friendly ping, we are scheduling a 1.1.0-M1 release soon, would you like to work on this this weekend, thanks. @GreyPlane

File: stream-tests/src/test/scala/org/apache/pekko/stream/DslConsistencySpec.scala (hunk 11 above)
Comment (He-Pin) [inline stream-tests/src/test/scala/org/apache/pekko/stream/DslConsistencySpec.scala:74]: remove it, no longer needed

File: project/StreamOperatorsIndexGenerator.scala [hunk 13]
```diff
@@ -111,7 +112,8 @@ object StreamOperatorsIndexGenerator extends AutoPlugin {
       "fromGraph",
       "actorSubscriber",
       "foldAsync",
-      "newOnCompleteStage"))
+      "newOnCompleteStage",
+      "exists"))
```
Comment (He-Pin) [inline project/StreamOperatorsIndexGenerator.scala:116]: I think this is not needed, as https://github.com/apache/incubator-pekko/pull/989/

File: stream/src/main/scala/org/apache/pekko/stream/javadsl/Source.scala [hunk 14]
```diff
@@ -3543,6 +3543,27 @@ final class Source[Out, Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[
...
+   *
+   * '''Emits when''' predicate returned true or upstream completes
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' predicate returned true or upstream completes
+   *
+   * '''Cancels when''' predicate returned true or downstream cancels
+   */
+  def exists(p: function.Predicate[Out]): Source[java.lang.Boolean, Mat] =
```
Comment (pjfanning) [inline stream/src/main/scala/org/apache/pekko/stream/javadsl/Source.scala:3564]: `@since 1.1.0` needed on al new public methods

Comment (GreyPlane) [reply stream/src/main/scala/org/apache/pekko/stream/javadsl/Source.scala:3565]: added

File: stream/src/main/scala/org/apache/pekko/stream/javadsl/Flow.scala [hunk 15]
```diff
@@ -1835,6 +1835,27 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
...
+   * '''Emits when''' predicate returned true or upstream completes
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' predicate returned true or upstream completes
+   *
+   * '''Cancels when''' predicate returned true or downstream cancels
+   */
+  def exists(p: function.Predicate[Out]): javadsl.Flow[In, java.lang.Boolean, Mat] =
+    new Flow(delegate.exists(p.test).map(Boolean.box))
```
Comment (He-Pin) [inline stream/src/main/scala/org/apache/pekko/stream/javadsl/Flow.scala:1858]: if we plan to add this to `flow/source`, then `forall` operator will need to do the same.
Currently, in ZIO, these method only exists on `ZSink`, but in fs2, these method exists on `Stream`.
Maybe we can just have these method on Sink for now too, wdyt, @GreyPlane @laglangyue @pjfanning @mdedetrich

File: stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala (hunk 3 above)
Comment (He-Pin) [reply stream/src/main/scala/org/apache/pekko/stream/scaladsl/Flow.scala:1550]: in the Flow/Source, the name attributes should placed in `DefaultAttributes`.
And you will need via( new  Fold...withAttributes(...)`.

**Comment by He-Pin** [comment]
I will take care of it, as I discussed with @GreyPlane.

**Review by jxnu-liguobin — APPROVED** [review]
lgtm

File: docs/src/main/paradox/stream/operators/Sink/forall.md [hunk 16]
```diff
@@ -44,4 +44,4 @@ Java
 
 **backpressures** when the invocation of predicate `p` has not yet completed
 
-@@@
\ No newline at end of file
+@@@
```
Comment (laglangyue) [inline docs/src/main/paradox/stream/operators/Sink/forall.md:47]: A small issue, seemingly unrelated change

**Comment by laglangyue** [comment]
LGTM, although has an irrelevant issue.

File: docs/src/main/paradox/stream/operators/Sink/forall.md (hunk 16 above)
Comment (He-Pin) [reply docs/src/main/paradox/stream/operators/Sink/forall.md:47]: Add a bland endline here.

**Review by Roiocam — APPROVED** [review]
LGTM, Thanks.

**Comment by He-Pin** [comment]
As I have rewritten most of the code, I think we can bring this in, and @GreyPlane Thank you for your time.
